validator/src/bootstrap.rs

```diff
@@ -1563,506 +1118,508 @@ mod with_incremental_snapshots {
+                Some((*base_snapshot_hash2, incremental_snapshot_hashes1.clone())),
+            ),
+        );
 
-            // handle duplicates, with different hashes
-            oracle.insert(
-                Pubkey::new_unique(),
-                (
-                    full_snapshot_hashes2.clone(),
-                    Some((*base_snapshot_hash2, incremental_snapshot_hashes2.clone())),
-                ),
-            );
+        // full and incremental snapshots, with different hashes
+        oracle.insert(
+            Pubkey::new_unique(),
+            (
+                full_snapshot_hashes2.clone(),
+                Some((*base_snapshot_hash2, incremental_snapshot_hashes2.clone())),
+            ),
+        );
 
-            let node_to_full_snapshot_hashes = |node| oracle.get(node).unwrap().clone().0;
-            let node_to_incremental_snapshot_hashes = |node| oracle.get(node).unwrap().clone().1;
+        // full and incremental snapshots, but base hash is wrong
+        oracle.insert(
+            Pubkey::new_unique(),
+            (
+                full_snapshot_hashes2.clone(),
+                Some((*base_snapshot_hash1, incremental_snapshot_hashes2.clone())),
+            ),
+        );
 
-            let known_snapshot_hashes = build_known_snapshot_hashes(
-                oracle.keys(),
-                node_to_full_snapshot_hashes,
-                node_to_incremental_snapshot_hashes,
-            );
+        // handle duplicates as well
+        oracle.insert(
+            Pubkey::new_unique(),
+            (
+                full_snapshot_hashes1.clone(),
+                Some((*base_snapshot_hash1, incremental_snapshot_hashes1.clone())),
+            ),
+        );
 
-            let mut known_full_snapshot_hashes: Vec<_> =
-                known_snapshot_hashes.keys().copied().collect();
-            known_full_snapshot_hashes.sort_unstable();
+        // handle duplicates, with different hashes
+        oracle.insert(
+            Pubkey::new_unique(),
+            (
+                full_snapshot_hashes2.clone(),
+                Some((*base_snapshot_hash2, incremental_snapshot_hashes2.clone())),
+            ),
+        );
 
-            let known_base_snapshot_hash = known_full_snapshot_hashes.last().unwrap();
+        let node_to_full_snapshot_hashes = |node| oracle.get(node).unwrap().clone().0;
+        let node_to_incremental_snapshot_hashes = |node| oracle.get(node).unwrap().clone().1;
 
-            let mut known_incremental_snapshot_hashes: Vec<_> = known_snapshot_hashes
-                .get(known_base_snapshot_hash)
-                .unwrap()
-                .iter()
-                .copied()
-                .collect();
-            known_incremental_snapshot_hashes.sort_unstable();
+        let known_snapshot_hashes = build_known_snapshot_hashes(
+            oracle.keys(),
+            node_to_full_snapshot_hashes,
+            node_to_incremental_snapshot_hashes,
+            true,
```

Comment:
Please update this test for when incremental snapshot fetch is `false`. Should be able to call `build_known_snapshot_hashes()` again with `false` and then assert with the expected snapshot hashes.